### ISSUE TITLE
Enhancement: Support tags or some other filtering mechanism

### ISSUE THREAD

**Issue by archaeogeek** [geopython/GeoHealthCheck#75]
I'd like to be able to define tags for entries and then be able to filter by those tags. An entry might have more than one tag. Either tags could be free text, entered by the user when they create an entry, or defined by the admin and then picked when creating an entry. It should then be possible to filter by those tags in the left-hand menu or using the search box.

**Comment by justb4** [comment]
Good suggestion! I see a relation with #18. I have been thinking in similar lines: one possibility using key/value pairs, ala OSM. But plain tags may be simpler to implement and could possibly tackle both #18 and #75.

**Comment by samperd** [comment]
I am OK with moving forward with "tags" that seems pretty flexible.  Currently we have little need for access controls. We try to be open by default :)

**Comment by tomkralidis** [comment]
First pass implemented in master (#83).  Feedback / testing would be much appreciated :)

**Comment by samperd** [comment]
@tomkralidis I tried an upgrade of my quite old GHC to test this out. I am now getting DB errors I think. Check out related issue (#84).

**Comment by justb4** [comment]
I also upgraded. Found no issues upgrading, using an existing `instance` (with PostgreSQL) and `config_site.py` of the master branch. This is what I did:

- pulled latest GitHub 
- `paver setup` (probably not required)
- `paver create`
- run `python GeoHealthCheck/app.py`

This created the new tables `tag` and `resource_tags` but kept my existing tables `resource`, `run` etc intact.  That is good! Running `python GeoHealthCheck/models.py run` ok.

In the GUI, what worked:

- on the left the new filter with tags is shown
- I could add a Resource and add tags to it
- I see the `tag` and `resource_tags` records added to the DB
- I see the new tags in the GUI tag filter section

What did not work:

- clicking on a tag went to `http://localhost:8000/?lang=en&tag=mytag` but still showed all Resources
- in edit Resource: adding tags to existing or new (that already had tags) Resources 
- in edit Resource: removing tags

A good start!

**Comment by justb4** [comment]
To add: `add.html` and `resource.html` could show a `select.. option` (combobox) `input` to select from all existing tags. Also `resource.html` has 
`<input type="text" value="{{ resource.tags2csv }}" data-role="tagsinput" disabled/>` which does not allow adding new tags by means of the `disabled` keyword.

**Comment by justb4** [comment]
As for clicking on a tag shows all Resources, I think  https://github.com/geopython/GeoHealthCheck/blob/master/GeoHealthCheck/views.py#L65

```
    if tag is not None:
        response['resources'] = models.Resource.query.filter(
            models.Tag.name.in_([tag])).all()
```

Returns all Resources, so this query may need adaptation.

**Comment by justb4** [comment]
Suggestion: on https://github.com/geopython/GeoHealthCheck/blob/master/GeoHealthCheck/models.py#L73
add autoincremented `identifier`, like:

```
resource_tags = DB.Table('resource_tags',
DB.Column('identifier', DB.Integer, primary_key=True, autoincrement=True),
                         DB.Column('tag_id', DB.Integer,
                                   DB.ForeignKey('tag.id')),
                         DB.Column('resource_identifier', DB.Integer,
                                   DB.ForeignKey('resource.identifier')))
```
At least in PostgreSQL this makes the table manageable in tools like `phpPGAdmin`.

**Comment by tomkralidis** [comment]
Thanks for the feedback!  Added the following fixes:

- homepage query by tag fixed
- edit resource tags implemented
- model identifier changes implemented

Todo:

- select via autocomplete from existing tags
- fix tag counts

**Comment by justb4** [comment]
Yes I can edit (add, remove) tags, even for existing Resources that had no tags. Two enhancements proposed:

- the Edit, Cancel, Save buttons are next to the Name field, they would be better below the Tags entry
- the Tags input field can be wider

**Comment by tomkralidis** [comment]
@justb4 thanks for the feedback.  I've moved the Edit, Cancel, Save buttons into their own row in the Resource page.  Where should the tags input field be wider? On the resource info page or the add resource page? Or both (changes welcome)?

**Comment by justb4** [comment]
@tomkralidis ok, good for the buttons, the tags input field is ok: only 2 enhancements directly seen:

- when editing tags in resource.html, no suggestions for existing tags
- after edit/refresh main dashboard, the tag counts in the tag-filter menu left are not updated

**Comment by justb4** [comment]
@tomkralidis working reviving/expanding the tests in `tests` dir, I found the following improvements for the `models.Tag` table:

- allow creation programmatically (e.g. from fixtures file)
- set constraints non-nullable and unique on `name` field (similar to `User.email`)

Results in this:

```
class Tag(DB.Model):
    id = DB.Column(DB.Integer, primary_key=True)
    name = DB.Column(DB.String(100), unique=True, nullable=False)

    def __init__(self, name):
        self.name = name
```

**Comment by tomkralidis** [comment]
Thanks @justb4; fixed.  master branch should now address the tagging functionality (suggestions, counts, editing, etc.).  If someone can test that would be great.

**Comment by justb4** [comment]
Tested, handling much smoother with `<option>` elements!, issues:

- adding tags on creating resources: adding an existing and new tag does not create the new tag
- removing a tag in Edit (sometimes) gives a duplicate key error in `/resource/<N>/update` and app needs restart

```
IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely) (psycopg2.IntegrityError) duplicate key value violates unique constraint "tag_name_key"
DETAIL:  Key (name)=(tag2) already exists.
 [SQL: 'INSERT INTO tag (name) VALUES (%(name)s) RETURNING tag.id'] [parameters: {'name': u'tag2'}]
```

Last error is hard to reproduce, several times seen. First the `update` fails and later I get the error screen.

**Comment by tomkralidis** [comment]
Thanks @justb4. first issue fixed, I cannot reproduce the second issue.

**Comment by justb4** [comment]
I'll look into second, after merging master into https://github.com/justb4/GeoHealthCheck/tree/plugins_82. As the Edit-part of `resource.html` is now in `edit_resource.html`, also moved the `JQuery select2`, appeared to be handy for `Probe/Check` editing as well!

**Comment by archaeogeek** [comment]
I tried to do an in-place upgrade to the latest version to include this functionality, and failed on the paver create step. In the database it created a tags table but not resource_tags, unfortunately I got no error, it just sat at the "Creating database objects" step for an hour. I got around this by creating a new database, editing config_site.py, running paver create and then restoring the old database over the top. So far so good, though I haven't done much tag creation yet...

**Comment by justb4** [comment]
My guess: you are using PostGIS. I found some hanging issues exclusively with PG, never SQLite. Found that SQLAlchemy is in cases [aggressively locking PG](http://stackoverflow.com/questions/6778942/python-sqlalchemy-postgresql-program-freezes). Explicitly closing SQLA `Session`s (after `commit()) helped. This is for the GHC code to be fixed. In some cases it happened when GHC Flask web app was running at the same time as the upgrade. Stopping GHC helped. But looks like you found a workaround.
